arch/risc-v/src/common/riscv_swint.c

```diff
@@ -440,6 +472,22 @@ int riscv_swint(int irq, void *context, void *arg)
           regs[REG_INT_CTX]   |= MSTATUS_MPPM; /* Machine mode */
 
           rtcb->xcp.sigreturn  = 0;
+
+#ifdef CONFIG_ARCH_KERNEL_STACK
+          /* We must enter here be using the user stack.  We need to switch
+           * to back to the kernel user stack before returning to the kernel
+           * mode signal trampoline.
+           */
+
+          if (rtcb->xcp.kstack != NULL)
+            {
+              DEBUGASSERT(rtcb->xcp.kstkptr != NULL &&
+                          (uintptr_t)rtcb->xcp.ustkptr == regs[REG_SP]);
```

Comment:
is it true?

Reply:
This is also copied from arch/arm/src/armv7-a/arm_syscall.c

I have not yet had time to test the userspace signal handler, so cannot verify if this works. To me it looks fine.

The comment at least is true, this syscall must come from userspace and the user stack must be in use.

Reply:
This first condition(rtcb->xcp.kstkptr != NULL) looks good, but the second((uintptr_t)rtcb->xcp.ustkptr == regs[REG_SP]) look strange.

Reply:
I assume the test is here to ensure that the user stack has been fully unwound by the userspace signal handler, i.e. when we return here the SP when dispatching the signal handler should be the same as when returning from the signal handler.

I cannot say for sure as I did not write this code.

Reply:
Ok, let keep until the assert trigger